File: Tellma/Entities/AccountType.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Tellma.Entities
{
    [StrongEntity]
    [EntityDisplay(Singular = "AccountType", Plural = "AccountTypes")]
    public class AccountTypeForSave<TCustodyDef, TResourceDef> : EntityWithKey<int>
    {
        [NotMapped]
        public int? ParentIndex { get; set; }

        [Display(Name = "TreeParent")]
        [AlwaysAccessible]
        [SelfReferencing(nameof(ParentIndex))]
        public int? ParentId { get; set; }

        [Display(Name = "Code")]
        [Required]
        [StringLength(50)]
        [AlwaysAccessible]
        public string Code { get; set; }

        [Display(Name = "AccountType_Concept")]
        [Required]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Concept { get; set; }

        [MultilingualDisplay(Name = "Name", Language = Language.Primary)]
        [Required]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name { get; set; }

        [MultilingualDisplay(Name = "Name", Language = Language.Secondary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name2 { get; set; }

        [MultilingualDisplay(Name = "Name", Language = Language.Ternary)]
        [StringLength(255)]
        [AlwaysAccessible]
        public string Name3 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Primary)]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Description { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Secondary)]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Description2 { get; set; }

        [MultilingualDisplay(Name = "Description", Language = Language.Ternary)]
        [StringLength(1024)]
        [AlwaysAccessible]
        public string Description3 { get; set; }

        [Display(Name = "AccountType_IsMonetary")]
        [AlwaysAccessible]
        public bool? IsMonetary { get; set; }

        [Display(Name = "IsAssignable")]
        [AlwaysAccessible]
        public bool? IsAssignable { get; set; }

        [Display(Name = "AccountType_StandardAndPure")]
        [AlwaysAccessible]
        public bool? StandardAndPure { get; set; }

        [Display(Name = "AccountType_CustodianDefinition")]
        [AlwaysAccessible]
        public int? CustodianDefinitionId { get; set; }

        [Display(Name = "AccountType_ParticipantDefinition")]
        [AlwaysAccessible]
        public int? ParticipantDefinitionId { get; set; }

        [Display(Name = "AccountType_EntryTypeParent")]
        [AlwaysAccessible]
        public int? EntryTypeParentId { get; set; }

        [MultilingualDisplay(Name = "AccountType_Time1Label", Language = Language.Primary)]
        [StringLength(50)]
        public string Time1Label { get; set; }

        [MultilingualDisplay(Name = "AccountType_Time1Label", Language = Language.Secondary)]
        [StringLength(50)]
        public string Time1Label2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_Time1Label", Language = Language.Ternary)]
        [StringLength(50)]
        public string Time1Label3 { get; set; }

        [MultilingualDisplay(Name = "AccountType_Time2Label", Language = Language.Primary)]
        [StringLength(50)]
        public string Time2Label { get; set; }

        [MultilingualDisplay(Name = "AccountType_Time2Label", Language = Language.Secondary)]
        [StringLength(50)]
        public string Time2Label2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_Time2Label", Language = Language.Ternary)]
        [StringLength(50)]
        public string Time2Label3 { get; set; }

        [MultilingualDisplay(Name = "AccountType_ExternalReferenceLabel", Language = Language.Primary)]
        [StringLength(50)]
        public string ExternalReferenceLabel { get; set; }

        [MultilingualDisplay(Name = "AccountType_ExternalReferenceLabel", Language = Language.Secondary)]
        [StringLength(50)]
        public string ExternalReferenceLabel2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_ExternalReferenceLabel", Language = Language.Ternary)]
        [StringLength(50)]
        public string ExternalReferenceLabel3 { get; set; }

        [MultilingualDisplay(Name = "AccountType_InternalReferenceLabel", Language = Language.Primary)]
        [StringLength(50)]
        public string InternalReferenceLabel { get; set; }

        [MultilingualDisplay(Name = "AccountType_InternalReferenceLabel", Language = Language.Secondary)]
        [StringLength(50)]
        public string InternalReferenceLabel2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_InternalReferenceLabel", Language = Language.Ternary)]
        [StringLength(50)]
        public string InternalReferenceLabel3 { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedAgentNameLabel", Language = Language.Primary)]
        [StringLength(50)]
        public string NotedAgentNameLabel { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedAgentNameLabel", Language = Language.Secondary)]
        [StringLength(50)]
        public string NotedAgentNameLabel2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedAgentNameLabel", Language = Language.Ternary)]
        [StringLength(50)]
        public string NotedAgentNameLabel3 { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedAmountLabel", Language = Language.Primary)]
        [StringLength(50)]
        public string NotedAmountLabel { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedAmountLabel", Language = Language.Secondary)]
        [StringLength(50)]
        public string NotedAmountLabel2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedAmountLabel", Language = Language.Ternary)]
        [StringLength(50)]
        public string NotedAmountLabel3 { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedDateLabel", Language = Language.Primary)]
        [StringLength(50)]
        public string NotedDateLabel { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedDateLabel", Language = Language.Secondary)]
        [StringLength(50)]
        public string NotedDateLabel2 { get; set; }

        [MultilingualDisplay(Name = "AccountType_NotedDateLabel", Language = Language.Ternary)]
        [StringLength(50)]
        public string NotedDateLabel3 { get; set; }

        [Display(Name = "AccountType_CustodyDefinitions")]
        [ForeignKey(nameof(AccountTypeCustodyDefinition.AccountTypeId))]
        public List<TCustodyDef> CustodyDefinitions { get; set; }

        [Display(Name = "AccountType_ResourceDefinitions")]
        [ForeignKey(nameof(AccountTypeResourceDefinition.AccountTypeId))]
        public List<TResourceDef> ResourceDefinitions { get; set; }
    }

    public class AccountTypeForSave : AccountTypeForSave<AccountTypeCustodyDefinitionForSave, AccountTypeResourceDefinitionForSave>
    {
    }

    public class AccountType : AccountTypeForSave<AccountTypeCustodyDefinition, AccountTypeResourceDefinition>
    {
        [AlwaysAccessible]
        public string Path { get; set; }

        [AlwaysAccessible]
        public short? Level { get; set; }

        [AlwaysAccessible]
        public int? ActiveChildCount { get; set; }

        [AlwaysAccessible]
        public int? ChildCount { get; set; }

        [Display(Name = "IsActive")]
        [AlwaysAccessible]
        public bool? IsActive { get; set; }

        [AlwaysAccessible]
        public bool? IsBusinessUnit { get; set; }

        [Display(Name = "IsSystem")]
        [AlwaysAccessible]
        public bool? IsSystem { get; set; }

        [Display(Name = "ModifiedBy")]
        public int? SavedById { get; set; }

        // For Query

        [AlwaysAccessible]
        public HierarchyId Node { get; set; }

        [Display(Name = "TreeParent")]
        [ForeignKey(nameof(ParentId))]
        public AccountType Parent { get; set; }

        [Display(Name = "ModifiedBy")]
        [ForeignKey(nameof(SavedById))]
        public User SavedBy { get; set; }

        [Display(Name = "AccountType_CustodianDefinition")]
        [ForeignKey(nameof(CustodianDefinitionId))]
        public RelationDefinition CustodianDefinition { get; set; }

        [Display(Name = "AccountType_ParticipantDefinition")]
        [ForeignKey(nameof(ParticipantDefinitionId))]
        public RelationDefinition ParticipantDefinition { get; set; }

        [Display(Name = "AccountType_EntryTypeParent")]
        [ForeignKey(nameof(EntryTypeParentId))]
        public EntryType EntryTypeParent { get; set; }
    }
}

File: Tellma/ClientApp/src/app/data/entities/account-type.ts
// tslint:disable:max-line-length
// tslint:disable:variable-name
import { EntityForSave } from './base/entity-for-save';
import { SettingsForClient } from '../dto/settings-for-client';
import { EntityDescriptor } from './base/metadata';
import { WorkspaceService } from '../workspace.service';
import { TranslateService } from '@ngx-translate/core';
import { EntityWithKey } from './base/entity-with-key';
import { DefinitionsForClient } from '../dto/definitions-for-client';
import { AccountTypeCustodyDefinitionForSave, AccountTypeCustodyDefinition } from './account-type-custody-definition';
import { AccountTypeResourceDefinitionForSave, AccountTypeResourceDefinition } from './account-type-resource-definition';

export type RequiredAssignment = 'A' | 'E';
export type OptionalAssignment = 'N' | 'A' | 'E';
export type EntryAssignment = 'N' | 'E';

export interface AccountTypeForSave<TCustodyDef = AccountTypeCustodyDefinitionForSave,
  TResourceDef = AccountTypeResourceDefinitionForSave> extends EntityForSave {
  ParentId?: number;
  Name?: string;
  Name2?: string;
  Name3?: string;
  Description?: string;
  Description2?: string;
  Description3?: string;
  Code?: string;
  Concept?: string;
  IsAssignable?: boolean;
  StandardAndPure?: boolean;
  CustodianDefinitionId?: number;
  ParticipantDefinitionId?: number;
  EntryTypeParentId?: number;
  NotedRelationDefinitionId?: number;
  IsMonetary?: boolean;
  Time1Label?: string;
  Time1Label2?: string;
  Time1Label3?: string;
  Time2Label?: string;
  Time2Label2?: string;
  Time2Label3?: string;
  ExternalReferenceLabel?: string;
  ExternalReferenceLabel2?: string;
  ExternalReferenceLabel3?: string;
  AdditionalReferenceLabel?: string;
  AdditionalReferenceLabel2?: string;
  AdditionalReferenceLabel3?: string;
  NotedAgentNameLabel?: string;
  NotedAgentNameLabel2?: string;
  NotedAgentNameLabel3?: string;
  NotedAmountLabel?: string;
  NotedAmountLabel2?: string;
  NotedAmountLabel3?: string;
  NotedDateLabel?: string;
  NotedDateLabel2?: string;
  NotedDateLabel3?: string;

  CustodyDefinitions?: TCustodyDef[];
  ResourceDefinitions?: TResourceDef[];
}

export interface AccountType extends AccountTypeForSave<AccountTypeCustodyDefinition, AccountTypeResourceDefinition> {
  Path?: string;
  Level?: number;
  ActiveChildCount?: number;
  ChildCount?: number;
  IsActive?: boolean;
  IsSystem?: boolean;
  SavedById?: number | string;

  CustodyDefinitionsCount?: number;
  ResourceDefinitionsCount?: number;
}

const _select = ['', '2', '3'].map(pf => 'Name' + pf);
let _settings: SettingsForClient;
let _definitions: DefinitionsForClient;
let _cache: EntityDescriptor = null;

export function metadata_AccountType(wss: WorkspaceService, trx: TranslateService): EntityDescriptor {
  const ws = wss.currentTenant;
  // Some global values affect the result, we check here if they have changed, otherwise we return the cached result
  if (ws.settings !== _settings || ws.definitions !== _definitions) {
    _settings = ws.settings;
    _definitions = ws.definitions;

    // clear the cache
    _cache = null;
  }

  if (!_cache) {
    const entityDesc: EntityDescriptor = {
      collection: 'AccountType',
      titleSingular: () => trx.instant('AccountType'),
      titlePlural: () => trx.instant('AccountTypes'),
      select: _select,
      apiEndpoint: 'account-types',
      masterScreenUrl: 'account-types',
      orderby: () => ws.isSecondaryLanguage ? [_select[1], _select[0]] : ws.isTernaryLanguage ? [_select[2], _select[0]] : [_select[0]],
      inactiveFilter: 'IsActive eq true',
      format: (item: EntityWithKey) => ws.getMultilingualValueImmediate(item, _select[0]),
      properties: {
        Id: { control: 'number', label: () => trx.instant('Id'), minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Name: { control: 'text', label: () => trx.instant('Name') + ws.primaryPostfix },
        Name2: { control: 'text', label: () => trx.instant('Name') + ws.secondaryPostfix },
        Name3: { control: 'text', label: () => trx.instant('Name') + ws.ternaryPostfix },
        Description: { control: 'text', label: () => trx.instant('Description') + ws.primaryPostfix },
        Description2: { control: 'text', label: () => trx.instant('Description') + ws.secondaryPostfix },
        Description3: { control: 'text', label: () => trx.instant('Description') + ws.ternaryPostfix },
        IsMonetary: { control: 'boolean', label: () => trx.instant('AccountType_IsMonetary') },
        Code: { control: 'text', label: () => trx.instant('Code') },
        Concept: { control: 'text', label: () => trx.instant('AccountType_Concept') },
        IsAssignable: { control: 'boolean', label: () => trx.instant('IsAssignable') },
        StandardAndPure: { control: 'boolean', label: () => trx.instant('AccountType_StandardAndPure') },
        CustodianDefinitionId: { control: 'number', label: () => `${trx.instant('AccountType_CustodianDefinition')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        CustodianDefinition: { control: 'navigation', label: () => trx.instant('AccountType_CustodianDefinition'), type: 'RelationDefinition', foreignKeyName: 'CustodianDefinitionId' },
        ParticipantDefinitionId: { control: 'number', label: () => `${trx.instant('AccountType_ParticipantDefinition')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        ParticipantDefinition: { control: 'navigation', label: () => trx.instant('AccountType_ParticipantDefinition'), type: 'RelationDefinition', foreignKeyName: 'ParticipantDefinitionId' },
        EntryTypeParentId: { control: 'number', label: () => `${trx.instant('AccountType_EntryTypeParent')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        EntryTypeParent: { control: 'navigation', label: () => trx.instant('AccountType_EntryTypeParent'), type: 'EntryType', foreignKeyName: 'EntryTypeParentId' },
        NotedRelationDefinitionId: { control: 'number', label: () => `${trx.instant('AccountType_NotedRelationDefinition')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        NotedRelationDefinition: { control: 'navigation', label: () => trx.instant('AccountType_NotedRelationDefinition'), type: 'RelationDefinition', foreignKeyName: 'NotedRelationDefinitionId' },
        Time1Label: { control: 'text', label: () => trx.instant('AccountType_Time1Label') + ws.primaryPostfix },
        Time1Label2: { control: 'text', label: () => trx.instant('AccountType_Time1Label') + ws.secondaryPostfix },
        Time1Label3: { control: 'text', label: () => trx.instant('AccountType_Time1Label') + ws.ternaryPostfix },
        Time2Label: { control: 'text', label: () => trx.instant('AccountType_Time2Label') + ws.primaryPostfix },
        Time2Label2: { control: 'text', label: () => trx.instant('AccountType_Time2Label') + ws.secondaryPostfix },
        Time2Label3: { control: 'text', label: () => trx.instant('AccountType_Time2Label') + ws.ternaryPostfix },
        ExternalReferenceLabel: { control: 'text', label: () => trx.instant('AccountType_ExternalReferenceLabel') + ws.primaryPostfix },
        ExternalReferenceLabel2: { control: 'text', label: () => trx.instant('AccountType_ExternalReferenceLabel') + ws.secondaryPostfix },
        ExternalReferenceLabel3: { control: 'text', label: () => trx.instant('AccountType_ExternalReferenceLabel') + ws.ternaryPostfix },
        AdditionalReferenceLabel: { control: 'text', label: () => trx.instant('AccountType_AdditionalReferenceLabel') + ws.primaryPostfix },
        AdditionalReferenceLabel2: { control: 'text', label: () => trx.instant('AccountType_AdditionalReferenceLabel') + ws.secondaryPostfix },
        AdditionalReferenceLabel3: { control: 'text', label: () => trx.instant('AccountType_AdditionalReferenceLabel') + ws.ternaryPostfix },
        NotedAgentNameLabel: { control: 'text', label: () => trx.instant('AccountType_NotedAgentNameLabel') + ws.primaryPostfix },
        NotedAgentNameLabel2: { control: 'text', label: () => trx.instant('AccountType_NotedAgentNameLabel') + ws.secondaryPostfix },
        NotedAgentNameLabel3: { control: 'text', label: () => trx.instant('AccountType_NotedAgentNameLabel') + ws.ternaryPostfix },
        NotedAmountLabel: { control: 'text', label: () => trx.instant('AccountType_NotedAmountLabel') + ws.primaryPostfix },
        NotedAmountLabel2: { control: 'text', label: () => trx.instant('AccountType_NotedAmountLabel') + ws.secondaryPostfix },
        NotedAmountLabel3: { control: 'text', label: () => trx.instant('AccountType_NotedAmountLabel') + ws.ternaryPostfix },
        NotedDateLabel: { control: 'text', label: () => trx.instant('AccountType_NotedDateLabel') + ws.primaryPostfix },
        NotedDateLabel2: { control: 'text', label: () => trx.instant('AccountType_NotedDateLabel') + ws.secondaryPostfix },
        NotedDateLabel3: { control: 'text', label: () => trx.instant('AccountType_NotedDateLabel') + ws.ternaryPostfix },

        // tree stuff
        Path: { control: 'text', label: () => trx.instant('TreePath') },
        ParentId: { control: 'number', label: () => `${trx.instant('TreeParent')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        Parent: { control: 'navigation', label: () => trx.instant('TreeParent'), type: 'AccountType', foreignKeyName: 'ParentId' },
        ChildCount: { control: 'number', label: () => trx.instant('TreeChildCount'), minDecimalPlaces: 0, maxDecimalPlaces: 0, alignment: 'right' },
        ActiveChildCount: { control: 'number', label: () => trx.instant('TreeActiveChildCount'), minDecimalPlaces: 0, maxDecimalPlaces: 0, alignment: 'right' },
        Level: { control: 'number', label: () => trx.instant('TreeLevel'), minDecimalPlaces: 0, maxDecimalPlaces: 0, alignment: 'right' },

        IsActive: { control: 'boolean', label: () => trx.instant('IsActive') },
        IsSystem: { control: 'boolean', label: () => trx.instant('IsSystem') },

        // Audit info
        SavedById: { control: 'number', label: () => `${trx.instant('ModifiedBy')} (${trx.instant('Id')})`, minDecimalPlaces: 0, maxDecimalPlaces: 0 },
        SavedBy: { control: 'navigation', label: () => trx.instant('ModifiedBy'), type: 'User', foreignKeyName: 'SavedById' }
      }
    };

    if (!ws.settings.SecondaryLanguageId) {
      delete entityDesc.properties.Name2;
      delete entityDesc.properties.Description2;
      delete entityDesc.properties.Time1Label2;
      delete entityDesc.properties.Time2Label2;
      delete entityDesc.properties.ExternalReferenceLabel2;
      delete entityDesc.properties.AdditionalReferenceLabel2;
      delete entityDesc.properties.NotedAgentNameLabel2;
      delete entityDesc.properties.NotedAmountLabel2;
      delete entityDesc.properties.NotedDateLabel2;
    }

    if (!ws.settings.TernaryLanguageId) {
      delete entityDesc.properties.Name3;
      delete entityDesc.properties.Description3;
      delete entityDesc.properties.Time1Label3;
      delete entityDesc.properties.Time2Label3;
      delete entityDesc.properties.ExternalReferenceLabel3;
      delete entityDesc.properties.AdditionalReferenceLabel3;
      delete entityDesc.properties.NotedAgentNameLabel3;
      delete entityDesc.properties.NotedAmountLabel3;
      delete entityDesc.properties.NotedDateLabel3;
    }

    _cache = entityDesc;
  }

  return _cache;
}
